### ISSUE TITLE
Full Rescan Interval not applied on startup

### ISSUE THREAD

**Issue by SkyLined** [syncthing/syncthing#5353]
I've setup a very long Full Rescan Interval for my folders but when I restart Syncthing, all folders are scanned even when they have already been scanned within this interval. If this is intentional, please add a setting to disable this.

Background: I am running Syncthing on a slow Raspberry Pi. Full Rescans consume 100% CPU, RAM and I/O, which makes the WebUI (and SSH) almost impossible to use. I am trying to avoid them by setting a very large interval but when I restart Syncthing, I basically cannot use the WebUI for a few hours because of all the unwanted rescans taking place.

### Version Information

Syncthing Version: v0.14.54, Linux (ARM)
OS Version: Linux 4.14.69+ on Raspberry Pi
Browser Version: n/a

**Comment by desbma** [comment]
> Full Rescans consume 100% CPU, RAM and I/O

Maybe you can improve that by lowering the CPU and IO priority, or even better use a different scheduler.

For example if you use the `syncthing@` service, with a `syncthing` user:
```
sudo mkdir -p /etc/systemd/system/syncthing@syncthing.service.d
echo '[Service]
CPUSchedulingPolicy=idle
IOSchedulingClass=idle' | sudo tee /etc/systemd/system/syncthing@syncthing.service.d/low-priority.conf
systemctl daemon-reload
systemctl restart syncthing@syncthing.service
```

For the other part of your issue, I am not familiar with that part of the code, but I suspect the initial scan can not be avoided, because Syncthing does not know if your files have been modified when it was not running.

**Comment by SkyLined** [comment]
Thanks, that suggestion sounds like it should solve some of my problems but might make the WebUI sluggish as it will also be throttled down... so, I'll give it a try to see how well it works.

Wrt. not knowning if a rescan is needed: in that case this bug would be a feature request for Syncthing to store a timestamp for the last scan on disk and read that at startup to avoid a rescan when none is needed.

**Comment by calmh** [comment]
Doing a scan at startup to kick off the schedule is intentional. With the default setting the difference isn't relevant. However in cases like yours where rescans should happen about once a day, I can see how rescanning everything on startup might be annoying. We could fix that, probably.

**Comment by SkyLined** [comment]
Fair enough, make this a feature request then :)

**Comment by desbma** [comment]
> Wrt. not knowning if a rescan is needed: in that case this bug would be a feature request for Syncthing to store a timestamp for the last scan on disk and read that at startup to avoid a rescan when none is needed.

How would you handle the case where the local files have been modified when the service is not running?

**Comment by desbma** [comment]
> Thanks, that suggestion sounds like it should solve some of my problems but might make the WebUI sluggish as it will also be throttled down...

If the scan is currently CPU bound, yet is likely to slow down the web UI too. However if is IO bound, it should remain reactive, only the background scan may take longer to run.

**Comment by calmh** [comment]
I think this could only apply in the case where filesystem notifications are disabled. If we are running with notifications we need the initial scan to have a baseline for notifications.

If we are running without notifications and have a rescan interval of an hour, it doesn't hurt to wait half an hour or whatever remains of the rescan interval on restart.

**Comment by SkyLined** [comment]
> How would you handle the case where the local files have been modified when the service (and the inotify watch) is not running?

Syncthing would now know about it until it did a full re-scan less than a day later. But this is not different from running Syncthing with "Watch for changes" disabled, so I do not see a problem with that (unless I am missing something).

**Comment by SkyLined** [comment]
> If we are running without notifications and have a rescan interval of an hour, it doesn't hurt to wait half an hour or whatever remains of the rescan interval on restart.

Indeed; in my case it is a day but I do not care if my backup is out of sync for a day.

**Comment by AudriusButkevicius** [comment]
I recall there were other reasons to run a scan. Mostly to prime caches todo with completion, but perhaps these days this is not important anymore.

**Comment by desbma** [comment]
Sorry if this has already been answered elsewhere, but on Linux when inotify is used (the default), why is a periodic rescan even needed?
If Syncthing first setups the inotify watches, and then does the initial scan (in that order to avoid races), why isn't that sufficient to be immediately notified of all futures filesystem changes?

**Comment by calmh** [comment]
Not all changes generate events, and events may be dropped due to overflow.

**Comment by desbma** [comment]
Care to elaborate?

I use inotify on another projet an I am interested in its limitations. I have only found some on FUSE filesystems.

If you are refering th the `/proc/sys/fs/inotify/max_*` overflow, then `syncthing` knows when it hits that limit, and could then fallback to heavyweight "full rescan to poll changes" approach (with a big fat warning to inform users that bumping the limit will improve performance).

**Comment by calmh** [comment]
IIRC, a chmod does not generate an event that we see.

I don't mean hitting max watches, which is handled as you say. I mean there being a flood of events faster than we can process them. There is best effort detection of this, but IIRC (again, I'm not super involved in the low level notify packages) we might not always know we missed events. Keep in mind there are a lot of underlying APIs (inotify, kqueue, whatever the thing is on Windows, etc) and the behavior of all is abstracted together.

**Comment by desbma** [comment]
> IIRC, a chmod does not generate an event that we see.

According to the inotify man page, it should:
```
IN_ATTRIB (*)
                  Metadata changed—for example, permissions (e.g.,
                  chmod(2)), timestamps (e.g., utimensat(2)), extended
                  attributes (setxattr(2)), link count (since Linux 2.6.25;
                  e.g., for the target of link(2) and for unlink(2)), and
                  user/group ID (e.g., chown(2)).
```
http://man7.org/linux/man-pages/man7/inotify.7.html

> I don't mean hitting max watches, which is handled as you say. I mean there being a flood of events faster than we can process them. There is best effort detection of this, but IIRC (again, I'm not super involved in the low level notify packages) we might not always know we missed events. 

You are probably refering to this code: https://github.com/syncthing/syncthing/blob/2f9840ddae38974684c25154352fde89bdd0e3e8/lib/fs/basicfs_watch.go#L76

If this is not sufficient to detect missed event, then I assume events can only be lost before that, in the [syncthing/notify](https://github.com/syncthing/notify) code.

> Keep in mind there are a lot of underlying APIs (inotify, kqueue, whatever the thing is on Windows, etc) and the behavior of all is abstracted together.

Yes, other APIs may have limitations, I am only interested in Linux + inotify.

> All in all, the notify thing is something like a 99% solution. It mostly does what you expect, but for correctness we should still do periodic scans now and then.

This still strikes me as suboptimal, but I am opening a new issue to discuss this instead of polluting this one.

**Comment by tsaridas** [comment]
I came across this post when searching on how to disable initial scanning.

My scan time can take days due to large amount of data and since I do scans through the API, when I try to scan a specific file/folder I have to wait for the initial scan to finish until I can scan for the one I need.

**Comment by Hoeze** [comment]
Are there any news on this? The initial scans eat 50% of my laptop battery overall...

**Comment by calmh** [comment]
The latest news, as of 12 seconds ago, is that we're not aware of anyone working on this. That could change at any time!

**Comment by AudriusButkevicius** [comment]
Or not!

**Comment by sir106** [comment]
would love if this is possible in the future - I have a NAS where changes are only possible through syncthing. would love if I could disable rescans on startup - my NAS is not up every day...

**Comment by dvcrn** [comment]
I would also love to see this in a future release. I'm running syncthing containerized and don't want a container restart to trigger a full scan (which takes a very very long time on my setup). It makes container and options tuning very hard because each restart would force rescan everything

**Comment by jms19** [comment]
Yes I'd like this. I have a one way sync for backup purposes and I don't run the receiver (backup machine) full time (disks, power consumption etc.) so given the presence of synching's persistent database doing a re-scan at start-up is a waste of resources.

**Comment by imsodin** [comment]
Please don't post "plus one" comments: https://github.com/syncthing/syncthing/wiki/NoPlusOne

**Comment by jms19** [comment]
When exactly is it a “plus one” comment as opposed to support of the issue stating one’s own reasoning ? This is something several people want yet the issue was closed.

**Comment by tomasz1986** [comment]
But this very issue is still open… 😉

Unless you mean a different one?

**Comment by fede843** [comment]
Hello, I have posted here: https://forum.syncthing.net/t/assume-unchanged-on-startup/12609/1

I am sharing my case here too, I think it is important to give some real cases where this feature can improve a scanario:

In my case, I have 2 USB disk attached to a RPI4. As you can imagine, patience is needed. It does the job perfectly, but when it comes to check for several TB at startup it gets lot of time reading (specially reading) and checking. So because of this, I need to think twice before restarting the OS. After a restart, the system spend a couple of hours, with 50% of CPU ussage (at least) and IO utilisation at 100% because of this, making the system quite unstable. If I had an option to disable this, I will schedule a cron job at 3am every day to do the scanning, that way I still get the changes in a daily basis (that is what I am interested for, in most of my cases) without affecting the system performance during active hours. I see lot of usage for this feature, many people requesting this here is under a similar situation.

**Comment by AudriusButkevicius** [comment]
I know this will be an unpopular and potentially offending opinion, and as quite a few comments here so far, add no substance to the issue, but people should really stop trying to feed terabytes of data through something akin to a smart calculator, and be shocked/surprised when it doesn't work well, and ask us to change syncthing.

**Comment by fede843** [comment]
Not offended, but we all don't have the same real options when it comes to buy HW. I am glad you have more powerful rig to sync files, I am.
I do have a RPI4 since several years ago, running lots of containers (homeassistant, plex, pihole, unbound, prometheus, grafana, loki, servarr apps, qbittorrent, etc), and I am really happy with it. It requires a lot of tuning, like this feature for instance.
It is easy to set up things when you have gigas and gigas of ram, horse power from several big x86 processors, etc etc. I still find some add value in trying to get things done in a such a small and wonderful device as a raspberry. To me, that is the best way of learning stuff, even with this calculator, as you misguidedly said before.
Having said that, there is no more need to brag nor deviate the request here. Keep your non constructive opinion yourself.

**Nevertheless, the request here still the same.**

**Comment by SkyLined** [comment]
Changing the hardware will not fix this issue. The problem is that the option to limit how often a rescan is done is not applied if you restart the application. Beefier hardware will just finish making this mistake quicker, not allow me to skip the scan when I told the application to do so.

**Comment by voruti** [comment]
A device with low power consumption is intended (in my use case).
The repeated scans are also disadvantageous in terms of energy efficiency (at least theoretically 😄).

**Comment by dgudim** [comment]
I am running syncthing on my phone and I've set it to start only on wifi, so every time I connect to wifi, all directories are getting rescanned, that's about 30k files, the phone kinda lags during this time. As far as I understand, there is still no option to disable this, is there?

**Comment by tomasz1986** [comment]
> As far as I understand, there is still no option to disable this, is there?

The issue is open, which usually means that the problem is still there 😉.

Honestly, if you've got that many files, on Android specifically, it may be better in terms of performance and battery life to just disable run conditions altogether and keep Syncthing running all the time.

**Comment by dgudim** [comment]
Well, yes, I thought there might be a workaround. If I run It without reatrictions it eats up my mobile data

**Comment by tomasz1986** [comment]
If you are tech-savvy and know what you are doing, you could use an automation app combined with `curl` (ex. via Termux) to control Syncthing using the API. This way, for example, you could only pause remote devices when disconnected from Wi-Fi and unpause them when reconnected to Wi-Fi again. This will save your mobile data without the need to restart Syncthing and take time to rescan the folders again and again. This method does require quite a bit of initial configuration to make everything work as supposed though.

I would suggest to open a topic on the forum if you want to go this route and need any help with it.

**Comment by Hoeze** [comment]
Is there a way to keep syncthing running and listen for file system changes but only distribute them once I am back on WiFi?

**Comment by isoos** [comment]
Question to syncthing developers: would you merge this feature if somebody implemented it / e.g. looking for contributors who implement it? Or is this something you wouldn't merge anyway?

**Comment by nvsystems** [comment]
Please allow this feature. The rescan on startup for large folders is incredible draining on the battery of laptops. And it's not useful: on a non-multi-boot system the files will not have changed between shutdown and the next boot. A periodic rescan occasionally is more than enough. Suggestion: something like "rescan after X days have elapsed (configurable) along with a checkbox [X] only when on mains power".

**Comment by Martchus** [comment]
> Question to syncthing developers: would you merge this feature if somebody implemented it / e.g. looking for contributors who implement it? Or is this something you wouldn't merge anyway?

I guess this comment from @calmh can be interpreted so that the answer to your last question is "yes":

> The latest news, as of 12 seconds ago, is that we're not aware of anyone working on this. That could change at any time!

And after briefly scrolling though the comments I also haven't read a clear "no" from the Syncthing devs who participated in the conversation.

---

> And it's not useful: on a non-multi-boot system the files will not have changed between shutdown and the next boot.

Files might have changed in the time window when Syncthing was not running. However, I agree that when keeping this time window low and when doing periodic rescans anyway than most people can probably live without a rescan on startup so it at least shouldn't be forced.

---

By the way, the usefulness of this feature came also up in https://github.com/Martchus/syncthingtray/issues/392 which is about dealing with metered network connections. One reason that speaks against completely stopping Syncthing when the network connection is metered are these forced rescans on startup. For many folders on my phone even manual rescans are good enough as they are *mostly* used as receive-only folders.

**Comment by rcarmo** [comment]
This (and being able to schedule syncs for specific times, not just intervals) would be amazing. I have a NAS where I can't set CPU or container priorities and where Syncthing completely swamps my CPU at random, and restarting it only makes the problem worse because it then goes and rescans everything...

**Comment by falcon4fun** [comment]
Similar problem here. Would ask to allow "Rescan on logon (once per XX hours at maximum)" like for example ESET Task Scheduler has "Run the task only once within the specified interval"

**Comment by milroneth** [comment]
If this hopefully gets implemented, I'd like to add that it would make a lot of sense to be a per-folder setting and not a global one. Small folders that see constant changes would better be re-scanned on startup, while larger less-active ones wouldn't.
